File: csharp/satellite/Satellite.cs
using System;
using System.Linq;

public record Tree(char Value, Tree? Left, Tree? Right);

public static class Satellite
{
    public static Tree? TreeFromTraversals(char[] preOrder, char[] inOrder)
    {
        if (preOrder.Length != inOrder.Length)
            throw new ArgumentException("traversals must have the same length");
        
        if (!preOrder.OrderBy(x => x).SequenceEqual(inOrder.OrderBy(x => x)))
            throw new ArgumentException("traversals must have the same elements");
        
        if (preOrder.Length != preOrder.Distinct().Count() || inOrder.Length != inOrder.Distinct().Count())
            throw new ArgumentException("traversals must contain unique items");

        return BuildTree(preOrder, inOrder);
    }

    private static Tree? BuildTree(char[] preOrder, char[] inOrder)
    {
        if (preOrder.Length == 0) return null;
        if (preOrder.Length == 1) return new Tree(preOrder[0], null, null);
        
        char root = preOrder[0];
        int rootIndex = Array.IndexOf(inOrder, root);

        var leftInOrder = inOrder.Take(rootIndex).ToArray();
        var rightInOrder = inOrder.Skip(rootIndex + 1).ToArray();

        var leftPreOrder = preOrder.Skip(1).Take(leftInOrder.Length).ToArray();
        var rightPreOrder = preOrder.Skip(1 + leftInOrder.Length).ToArray();

        return new Tree(root, BuildTree(leftPreOrder, leftInOrder), BuildTree(rightPreOrder, rightInOrder));
    }
}

File: fsharp/satellite/Satellite.fs
module Satellite

type Tree =
    | Empty
    | Node of value: string * left: Tree * right: Tree

let private splitAtElement el a = (List.takeWhile ((<>) el) a, List.skipWhile ((<>) el) a |> List.tail)

let treeFromTraversals inorder preorder : Result<Tree, string> =
    let rec loop io po : Tree = 
        match po with 
            | [] -> Empty 
            | [a] -> Node(a, Empty, Empty)
            | root :: rest -> 
                let (lft_io, rgt_io) = io |> splitAtElement root
                let (lft_po, rgt_po) = rest |> List.splitAt (lft_io |> List.length)
                Node(root, loop lft_io lft_po, loop rgt_io rgt_po)

    if   List.length inorder <> List.length preorder then Error "traversals must have the same length"
    elif Set.ofList inorder <> Set.ofList preorder then Error "traversals must have the same elements"
    elif inorder <> List.distinct inorder || preorder <> List.distinct preorder then Error "traversals must contain unique items"
    else Ok (loop inorder preorder)
